plugins/steelseries/fu-steelseries-device.c

```diff
@@ -46,51 +58,141 @@ fu_steelseries_device_setup(FuDevice *device, GError **error)
 		g_prefix_error(error, "failed to do control transfer: ");
 		return FALSE;
 	}
-	if (actual_len != 32) {
+	if (actual_len != STEELSERIES_BUFFER_CONTROL_SIZE) {
 		g_set_error(error,
 			    G_IO_ERROR,
 			    G_IO_ERROR_INVALID_DATA,
 			    "only wrote %" G_GSIZE_FORMAT "bytes",
 			    actual_len);
 		return FALSE;
 	}
+
+	/* cleanup the buffer before receiving any data */
+	memset(data, 0x00, STEELSERIES_BUFFER_CONTROL_SIZE);
+
+	/* do not expect the answer from device */
+	if (answer != TRUE)
+		return TRUE;
+
 	ret = g_usb_device_interrupt_transfer(usb_device,
-					      0x81, /* EP1 IN */
+					      priv->ep,
 					      data,
-					      sizeof(data),
+					      priv->ep_in_size,
 					      &actual_len,
 					      STEELSERIES_TRANSACTION_TIMEOUT,
 					      NULL,
 					      error);
 	if (!ret) {
-		g_prefix_error(error, "failed to do EP1 transfer: ");
+		g_prefix_error(error, "failed to do EP transfer: ");
 		return FALSE;
 	}
-	if (actual_len != 32) {
+	if (actual_len != priv->ep_in_size) {
 		g_set_error(error,
 			    G_IO_ERROR,
 			    G_IO_ERROR_INVALID_DATA,
 			    "only read %" G_GSIZE_FORMAT "bytes",
 			    actual_len);
 		return FALSE;
 	}
-	version = g_strdup_printf("%i.%i.%i", data[0], data[1], data[2]);
-	fu_device_set_version(FU_DEVICE(device), version);
 
 	/* success */
 	return TRUE;
 }
 
+static gboolean
+fu_steelseries_device_probe(FuDevice *device, GError **error)
+{
+#if G_USB_CHECK_VERSION(0, 3, 3)
+	FuSteelseriesDevice *self = FU_STEELSERIES_DEVICE(device);
+	FuSteelseriesDevicePrivate *priv = GET_PRIVATE(self);
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(device));
+	GUsbInterface *iface = NULL;
+	GUsbEndpoint *ep = NULL;
+	guint8 ep_id;
+	guint16 packet_size;
+	g_autoptr(GPtrArray) ifaces = NULL;
+	g_autoptr(GPtrArray) endpoints = NULL;
+
+	/* FuUsbDevice->probe */
+	if (!FU_DEVICE_CLASS(fu_steelseries_device_parent_class)->probe(device, error))
+		return FALSE;
+
+	ifaces = g_usb_device_get_interfaces(usb_device, error);
+	if (ifaces == NULL)
+		return FALSE;
+
+	/* use the correct interface for interrupt transfer, either specifying an absolute offset,
+	 * or a negative offset value for the "last" one */
+	if (priv->iface_idx_offset >= 0) {
+		if ((guint)priv->iface_idx_offset > ifaces->len) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_FOUND,
+				    "update interface 0x%x not found",
+				    priv->iface_idx_offset);
```

Comment:
Need to be casted to `guint` or use `%d`; I do not know which is the most appropriated.

Reply:
guint pls